### PR TITLE
Implement igamma/igammac backward w.r.t. first argument

### PR DESCRIPTION
**Pull request by jashshah999**
## Summary

Implements the gradient of `torch.special.gammainc` / `torch.special.gammaincc` (igamma/igammac) with respect to the first argument `a` (shape parameter). Previously calling `.backward()` on these functions raised `not_implemented("igamma: input")`.

- Uses two computation regimes based on Stan math library: asymptotic expansion (DLMF 8.11.2) when `z >= 8 && z >= a`, series expansion (DLMF 8.7.3 / Gautschi 5.4) otherwise
- Registers `igamma_self_backward` and `igammac_self_backward` as structured binary ops with CPU + CUDA dispatch
- Enables differentiable CDFs for Gamma/Poisson distributions (survival analysis, temporal point processes)

Based on the work by @drewfustin in #130782, rebased onto current main.

Fixes #80025

## Files changed (11)

**Core math** (`aten/src/ATen/native/Math.h`): `calc_igamma_grada`, `calc_igammac_grada` and helpers. Also adds `C10_HOST_DEVICE` to `calc_digamma` for CUDA compatibility.

**Op registration**: `native_functions.yaml`, `BinaryOps.cpp/h`, `NamedRegistrations.cpp`, `BatchRulesBinaryOps.cpp`

**Kernels**: `cpu/BinaryOpsKernel.cpp` (CPU), `cuda/IGammaBackwardKernel.cu` (CUDA, new file)

**Autograd**: `derivatives.yaml` - replaces `not_implemented` with `at::igamma_self_backward`

**Tests**: `common_methods_invocations.py` - enables `supports_autograd=True`, sets `supports_gradgrad=False`

**License**: `NOTICE` - Stan BSD 3-Clause attribution

## Test plan

- [ ] CI passes (codegen, CPU tests, CUDA tests)
- [ ] `pytest test/test_ops.py -vk igamma` -- OpInfo gradient tests
- [ ] `pytest test/functorch/test_ops.py -vk igamma` -- functorch tests
- [ ] Smoke test:
  \`\`\`python
  import torch
  a = torch.tensor(2.0, requires_grad=True)
  x = torch.tensor(0.5)
  r = torch.special.gammainc(a, x)
  r.backward()
  print(a.grad)  # should print a finite gradient
  \`\`\`

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10 @jerryzh168 @aditew01 @lezcano @peterbell10

### CODE REVIEW

File: aten/src/ATen/native/Math.h [hunk 1]
```diff
@@ -1230,6 +1231,205 @@ template <>
   return calc_igammac<float>(float(a), float(x));
 }
 
+// Backward pass for igamma/igammac w.r.t. first argument (a).
+// Derived from Stan math library's grad_reg_inc_gamma.
+// See note [Stan's 3-Clause BSD License] in NOTICE.
+
+template <typename opmath_t>
+static C10_HOST_DEVICE opmath_t
+_igammac_grada_helper_asymptotic_series(opmath_t a, opmath_t x) {
```
Comment (Skylion007) [inline aten/src/ATen/native/Math.h:1240]: Why don't we just implement this using FunctionsManual.cpp using native TensorOps? is there a major speed advantage here?

**Comment by jashshah999** [comment]
Re: @Skylion007's question about FunctionsManual.cpp --

The gradient of igamma w.r.t. `a` requires per-element iterative series expansions (up to 2000 terms with adaptive convergence) that branch between asymptotic and series regimes depending on relative magnitudes of `a` and `x`. There is no closed-form expression in terms of existing tensor ops.

This is the same situation as the forward pass -- `calc_igamma`/`calc_igammac` are pointwise kernels in Math.h for the same reason. A FunctionsManual.cpp approach would need to either loop in host code (no GPU support) or pre-allocate max iterations and mask out converged elements (wasteful, no early termination, and the 3-way branching makes it very awkward to vectorize).

The kernel approach gives CUDA support via `C10_HOST_DEVICE` + `gpu_kernel`, matching the forward pass pattern. TensorFlow and JAX use the same per-element kernel approach for this gradient.
